### ISSUE TITLE
Emit diagnostic when casting an object to a type disjoint from its inferred type

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ty#3702]
basedpyright's [`reportInvalidCast`](https://docs.basedpyright.com/latest/benefits-over-pyright/new-diagnostic-rules/#reportinvalidcast) rule detects casts that are likely a mistake by validating that the types overlap:

```py
from typing import cast


cast(str, 1)
```

```
Conversion of type `Literal[1]` to type `str` may be a mistake because neither type sufficiently overlaps with the other.  If this was intentional, convert the expression to `object` first.
```

[(typescript also does the same thing)](https://www.typescriptlang.org/play/?#code/IwAghgziEC4E4EsB2BzIA)

it would be nice if ty had a rule for this as well

**Comment by DetachHead** [comment]
i see this was addressed in https://github.com/astral-sh/ruff/pull/28129
